File: apps/api/app/routers/predictions.py
"""
Predictions router for ML-powered F1 predictions.
"""

from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query

from app.models.predictions import (
    PitPredictionRequest,
    PitPredictionResponse,
    PositionForecastResponse,
    StrategyAnalysisResponse,
)

router = APIRouter()


@router.post("/pit-stop", response_model=PitPredictionResponse)
async def predict_pit_stop(request: PitPredictionRequest) -> dict:
    """
    Predict the likelihood and timing of a pit stop.

    Uses ML model trained on historical telemetry and strategy data.
    """
    # TODO: Implement actual ML prediction
    # This is a placeholder response
    return {
        "driver_number": request.driver_number,
        "probability": 0.75,
        "predicted_lap": request.current_lap + 5,
        "confidence": 0.82,
        "reasons": [
            "Tyre degradation above threshold",
            "Traffic ahead may create opportunity",
            "Team historical strategy pattern",
        ],
    }


@router.get("/position-forecast", response_model=List[PositionForecastResponse])
async def get_position_forecast(
    session_key: int = Query(..., description="Session key"),
    driver_number: Optional[int] = Query(None, description="Driver number filter"),
    laps_ahead: int = Query(10, ge=1, le=20, description="Laps to forecast"),
) -> List[dict]:
    """
    Forecast driver positions for upcoming laps.

    Uses ML model to predict position changes based on pace, tyre life, and gaps.
    """
    # TODO: Implement actual ML prediction
    # This is a placeholder response
    drivers = [1, 11, 16, 44, 55, 63, 14, 31, 10, 22, 4, 27, 20, 21, 18, 77, 24, 81, 3, 23]

    if driver_number:
        drivers = [driver_number]

    return [
        {
            "driver_number": d,
            "current_position": idx + 1,
            "predicted_position": max(1, idx + 1 - (idx % 2)),
            "position_change": -(idx % 2),
            "confidence": 0.75 - (idx * 0.02),
        }
        for idx, d in enumerate(drivers)
    ]


@router.get("/strategy", response_model=StrategyAnalysisResponse)
async def analyze_strategy(
    session_key: int = Query(..., description="Session key"),
    driver_number: int = Query(..., description="Driver number"),
) -> dict:
    """
    Analyze and recommend pit stop strategy.

    Returns tyre strategy recommendations and risk assessment.
    """
    # TODO: Implement actual strategy analysis
    # This is a placeholder response
    return {
        "driver_number": driver_number,
        "current_tyre": "MEDIUM",
        "tyre_age": 15,
        "recommended_stops": 1,
        "optimal_lap_window": [18, 25],
        "risk_level": "medium",
    }

File: packages/shared/src/types/predictions.ts
// Prediction types shared between frontend and backend

export interface PitPrediction {
  driver_number: number;
  probability: number;
  predicted_lap: number;
  recommended_compound?: string;
  confidence: number;
  reasons: string[];
}

export interface DriverPositionForecast {
  driver_number: number;
  driver_name?: string;
  team_name?: string;
  current_position: number;
  predicted_position: number;
  position_change: number;
  confidence: number;
  factors?: string[];
}

export interface PositionForecastResponse {
  session_key: number;
  current_lap?: number;
  total_laps?: number;
  predictions: DriverPositionForecast[];
  generated_at?: string;
}

export interface StrategyAlternative {
  strategy: string;
  risk: string;
  expected_gain?: number;
  condition?: string;
}

export interface StrategyAnalysis {
  driver_number: number;
  current_compound: string;
  tyre_age: number;
  recommended_stops: number;
  optimal_laps: number[];
  compounds: string[];
  risk_level: 'low' | 'medium' | 'high';
  expected_positions_gained: number;
  confidence: number;
  alternative_strategies: StrategyAlternative[];
  factors: string[];
}

export interface AllPredictionsResponse {
  session_key: number;
  pit_predictions: PitPrediction[];
  position_forecast: PositionForecastResponse;
  generated_at: string;
}
